heft-plugins/heft-typescript-plugin/src/TypeScriptPlugin.ts

```diff
@@ -0,0 +1,390 @@
+      // all source files to this set of static assets. This would allow us to avoid
+      // having to copy the static assets multiple times, increasing build times and
+      // package size.
+      await this._updateStaticAssetsToCopy(taskSession, heftConfiguration, runOptions);
+    });
+  }
+
+  private async _updateClean(
+    taskSession: IHeftTaskSession,
+    heftConfiguration: HeftConfiguration,
+    cleanOptions: IHeftTaskCleanHookOptions
+  ): Promise<void> {
+    const configurationFile: ITypeScriptConfigurationJson | undefined =
+      await loadTypeScriptConfigurationFileAsync(heftConfiguration, taskSession.logger.terminal);
+
+    // For now, delete the entire output folder and additional module kind output folders. In the future,
+    // we may want to clean specific files that we know are produced by the TypeScript compiler.
+    const tsconfigOutDir: string | undefined = await this._getTsconfigOutDirAsync(
+      taskSession,
+      heftConfiguration,
+      configurationFile
+    );
+    if (tsconfigOutDir) {
+      cleanOptions.addDeleteOperations({ sourceFolder: tsconfigOutDir });
+    }
+
+    if (configurationFile?.additionalModuleKindsToEmit) {
+      for (const additionalModuleKindToEmit of configurationFile.additionalModuleKindsToEmit) {
+        cleanOptions.addDeleteOperations({
+          sourceFolder: path.resolve(heftConfiguration.buildFolder, additionalModuleKindToEmit.outFolderName)
+        });
+      }
+    }
+  }
+
+  private async _updateStaticAssetsToCopy(
+    taskSession: IHeftTaskSession,
+    heftConfiguration: HeftConfiguration,
+    runOptions: IHeftTaskRunHookOptions
+  ): Promise<void> {
+    const typeScriptConfiguration: ITypeScriptConfigurationJson | undefined =
+      await loadTypeScriptConfigurationFileAsync(heftConfiguration, taskSession.logger.terminal);
+
+    // We only care about the copy if static assets were specified.
+    if (
+      typeScriptConfiguration?.staticAssetsToCopy?.fileExtensions?.length ||
+      typeScriptConfiguration?.staticAssetsToCopy?.includeGlobs?.length ||
+      typeScriptConfiguration?.staticAssetsToCopy?.excludeGlobs?.length
+    ) {
+      const destinationFolderPaths: Set<string> = new Set<string>();
+
+      // Add the output folder and all additional module kind output folders as destinations
+      const tsconfigOutDir: string | undefined = await this._getTsconfigOutDirAsync(
+        taskSession,
+        heftConfiguration,
+        typeScriptConfiguration
+      );
+      if (tsconfigOutDir) {
+        destinationFolderPaths.add(tsconfigOutDir);
+      }
+      for (const emitModule of typeScriptConfiguration?.additionalModuleKindsToEmit || []) {
+        destinationFolderPaths.add(path.resolve(heftConfiguration.buildFolder, emitModule.outFolderName));
+      }
+
+      runOptions.addCopyOperations({
+        ...typeScriptConfiguration?.staticAssetsToCopy,
+
+        // For now - these may need to be revised later
+        sourceFolder: 'src',
```

Comment:
Feels like we should resolve this from the tsconfig `rootDir` at this point.

Reply:
Possibly? But we should probably do that in a separate PR.